core/domain/classifier_services.py

```diff
@@ -424,41 +375,240 @@ def delete_classifier_training_job(job_id):
+    algorithm_id, or if the job_id exists but it has been trained on a now
+    obsolete algorithm, we re-submit the jobs.
+
+    The function goes through existing training job exploration mapping and
+    identifies three different types of algorithm IDs.
+        1. algorithm_ids_to_upgrade: Those which exist but needs to be
+            upgraded a new algorithm (because existing one has been deprecated)
+            by re-submitting the training job.
+        2. algorithm_ids_to_add: Those which doesn't exist and needs to be added
+            by submitting a new training job.
+        3. algorithm_ids_to_remove: Those which needs to be removed since these
+            algorithms are no longer supported.
+
+    Once all three types of algorithm IDs are filtered, the function performs
+    specific tasks tailored to each of them. Its called lazy migration because
+    it happens only when there is a query to retrieve trained model for given
+    <exploration, version, state> and algorithm_id.
+
+    Args:
+        state_training_jobs_mapping: StateTrainingJobsMapping. Domain
+            object containing exploration to training job id mapping. This
+            mapping is used to figure out jobs that needs to be re-submitted /
+            added or removed.
+    """
+    exp_id = state_training_jobs_mapping.exp_id
+    exp_version = state_training_jobs_mapping.exp_version
+    state_name = state_training_jobs_mapping.state_name
+
+    exploration = exp_fetchers.get_exploration_by_id(
+        exp_id, version=exp_version)
+    interaction_id = exploration.states[state_name].interaction.id
+
+    algorithm_id = feconf.INTERACTION_CLASSIFIER_MAPPING[
+        interaction_id]['algorithm_id']
+    algorithm_version = feconf.INTERACTION_CLASSIFIER_MAPPING[
+        interaction_id]['algorithm_version']
+
+    algorithm_id_to_algorithm_version = {
+        algorithm_id: algorithm_version
+    }
+    possible_algorithm_ids = [algorithm_id]
+
+    algorithm_ids_to_add = [
+        algorithm_id for algorithm_id in possible_algorithm_ids
+        if not (
+            state_training_jobs_mapping.algorithm_ids_to_job_ids.
+            contains_algorithm_id(algorithm_id))]
+
+    algorithm_ids_to_remove = [
+        algorithm_id for algorithm_id in (
+            state_training_jobs_mapping.algorithm_ids_to_job_ids.
+            get_all_algorithm_ids())
+        if algorithm_id not in possible_algorithm_ids]
+
+    algorithm_ids_to_upgrade = [
+        algorithm_id for algorithm_id in possible_algorithm_ids
+        if (state_training_jobs_mapping.algorithm_ids_to_job_ids.
+            contains_algorithm_id(algorithm_id))]
+
+    if len(algorithm_ids_to_add) > 0:
+        job_dicts_list = []
+
+        for algorithm_id in algorithm_ids_to_add:
+            next_scheduled_check_time = datetime.datetime.utcnow()
+            training_data = exploration.states[state_name].get_training_data()
+
+            dummy_classifier_training_job = (
```

Comment:
Let's not give this a name -- it makes it significant. Perhaps just do:

```
classifier_domain.ClassifierTrainingJob(
    ...
    ...
).validate()
```

Reply:
Done